### PR TITLE
Registration

### LINKED ISSUE #14: Registration Page
When a user visits the '/registration' path they should see a form to register.
The form should include:
- [ ] Email
- [ ] Password
- [ ] Password Confirmation
- [ ] Register Button

Once the user registers they should be logged in and redirected to the dashboard page

### PR DESCRIPTION
**Pull request by priyapower**
**Contributors:** Priya

Closes #14 

**Details:**
- Creates registration link from Welcome Page (and adds login form)
- Create initial registration form page
- Begins initial setup of creating a sessions controller

**Improvements:** (issues/bugs/refactoring?)

- SimpleCov (%): 30 / 45 LOC (66.67%)
- SimpleCov Model (%): 14 / 55 LOC (25.45%)
- SimpleCov Feature (%): 30 / 45 LOC (66.67%)
- RuboCop (Complaints/Violations): 11 files inspected, 12 offenses detected
Offenses:

app/controllers/users_controller.rb:2:3: C: Style/EmptyMethod: Put empty method definitions on a single line.
  def new ...
  ^^^^^^^
app/controllers/users_controller.rb:3:1: C: Layout/EmptyLinesAroundMethodBody: Extra empty line detected at method body beginning.
app/controllers/users_controller.rb:10:19: C: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      redirect_to "/dashboard"
                  ^^^^^^^^^^^^
app/controllers/users_controller.rb:12:25: C: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      flash[:failure] = "Something went wrong! Are you missing a required field? Are you already registered?"
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app/controllers/users_controller.rb:14:7: C: Layout/LeadingCommentSpace: Missing space after #.
      #Example: "ActiveRecord::RecordInvalid: Validation failed: Password confirmation doesn't match Password"
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app/controllers/users_controller.rb:22:3: C: Layout/EmptyLinesAroundAccessModifier: Keep a blank line before and after private.
  private
  ^^^^^^^
app/controllers/welcome_controller.rb:2:3: C: Style/EmptyMethod: Put empty method definitions on a single line.
  def index ...
  ^^^^^^^^^
app/models/user.rb:3:3: C: Rails/Validation: Prefer the new style validations validates :column, confirmation: value over validates_confirmation_of.
  validates_confirmation_of :password, require: true
  ^^^^^^^^^^^^^^^^^^^^^^^^^
app/models/user.rb:7:3: C: Rails/HasManyOrHasOneDependent: Specify a :dependent option.
  has_many :followed_users, foreign_key: :follower_id, class_name: 'Friendship'
  ^^^^^^^^
app/models/user.rb:7:3: C: Rails/InverseOf: Specify an :inverse_of option.
  has_many :followed_users, foreign_key: :follower_id, class_name: 'Friendship'
  ^^^^^^^^
app/models/user.rb:9:3: C: Rails/HasManyOrHasOneDependent: Specify a :dependent option.
  has_many :following_users, foreign_key: :followee_id, class_name: 'Friendship'
  ^^^^^^^^
app/models/user.rb:9:3: C: Rails/InverseOf: Specify an :inverse_of option.
  has_many :following_users, foreign_key: :followee_id, class_name: 'Friendship'
  ^^^^^^^^

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    end
+  end
+
+  def show
+    @user = User.find(session[:user_id])
+  end
+
+  private
+  def user_params
+    params.permit(:email, :password, :password_confirmation)
```
Comment (Arique1104) [inline app/controllers/users_controller.rb]: Do we need to permit `password_confirmation`?  I thought that adding a password confirmation was more for the user or Bcrypt gem, and less for us in our database validations.

File: app/models/user.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 class User < ApplicationRecord
   validates :email, uniqueness: true, presence: true
-  validates_presence_of :password, require: true
+  validates_confirmation_of :password, require: true
```
Comment (Arique1104) [inline app/models/user.rb:3]: This seems awesome!  Would love to hear more about this command.

File: app/views/users/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<!-- This needs to be centered and very large font -->
```
Comment (Arique1104) [inline app/views/users/new.html.erb:1]: I love these commented out notes!

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<!-- Wireframe shows an emoji or image in the top left corner -->
+<!-- The welcome is centered off the image -->
```
Comment (Arique1104) [inline app/views/users/show.html.erb:2]: Those wireframe works are sure helpful!

File: app/views/welcome/index.html.erb [hunk 5]
```diff
@@ -1 +1,19 @@
...
+
+<!-- This needs to be centered under the description -->
+<%= form_tag "/dashboard" do %>
+  <%= label_tag :email %>
+  <%= email_field_tag :email %>
+
+  <%= label_tag :password %>
+  <%= password_field_tag :password %>
+
+  <%= submit_tag 'Sign In', class: :button %>
```
Comment (Arique1104) [inline app/views/welcome/index.html.erb:15]: Nice form!  And taking advantage of the automated method to post from the `submit_tag`

File: config/routes.rb [hunk 6]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   get "/", to: 'welcome#index', as: :welcome
   # resources :welcome, :path => '/', only: :index, as: :welcome
-
+  get "/register", to: "users#new"
+  post "/users", to: "users#create"
+  get "/dashboard", to: "users#show"
```
Comment (Arique1104) [inline config/routes.rb:6]: I smell a name spacing coming on!  What do you think we should do when integrating authentication and authorization of our users and their API key.  I saw some documentation about how to load things to a session, I'm going to be really excited to figure out that next step.

Comment (Arique1104) [reply config/routes.rb:6]: Per our conversation: We don't have to deal with authorization, but we might need to consider simply our authentication process and no one who is authenticated can see anything.

File: spec/features/registration/registration_spec.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    fill_in :email, with: email
+    fill_in :password, with: password
+    fill_in :password_confirmation, with: password
+
+    click_button "Register"
+    expect(current_path).to eq("/dashboard")
+    expect(page).to have_content("Welcome #{email}!")
+  end
+
+  it "can give a failure if user is already in system"
```
Comment (Arique1104) [inline spec/features/registration/registration_spec.rb:58]: #SAD Path!  Very nice!

**Review by Arique1104 — CHANGES_REQUESTED** [review]
This looks great!  I think the only thing I suggested was not needing to permit password_confirmation because that's less important information that we need to be tracking and more to make sure that our user and our Bcrypt gem have all the info they need.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (priyapower) [reply app/controllers/users_controller.rb]: I believe this is to allow from the registration process
- A user goes to register
- They input email, password, and confirmation
- If their password does not match password confirmation, it sends back an error

Is this incorrect? I can try running it without, but I believe I have a sad path test for password and confirmation match.

File: app/views/welcome/index.html.erb (hunk 5 above)
Comment (priyapower) [reply app/views/welcome/index.html.erb:15]: Thanks!

File: config/routes.rb (hunk 6 above)
Comment (priyapower) [reply config/routes.rb:6]: Sessions are in the Welcome Branch! I'm glad to see you predicting that! Confirms that my methodology is matching yours :)

Comment (priyapower) [reply config/routes.rb:6]: We absolutely need to write in tests for logged out users getting 400 errors :)

File: app/models/user.rb (hunk 2 above)
Comment (priyapower) [reply app/models/user.rb:3]: I am going to check on your previous comment and see about the suspicions for password_confirmation before I can explain this :)

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (priyapower) [reply app/controllers/users_controller.rb]: This WAS extraneous! Removed and pushed up changes

**Review by Arique1104 — APPROVED** [review]
Great!

**Pull request merged**